### ISSUE TITLE
This plugin need to use the NUSPEC file

### ISSUE THREAD

**Issue by tebeco** [jfrog/artifactory-user-plugins#234]
There's a huge issue as this pluggin seems to base everything on the Filename ....

If I send you a mail with a file attached named `notepad.exe`
do you trust the file by the name and run it ?
or do you need to checked that it was Signed by MSFT + check that the certificate used to signed it still valid + the hash ?
Or you just delete the mail

Obviously for a repository manager you'll choose security over filename to avoid both spoofing attack  tampering right ?

When this plugin is applied on an existing repository, the result of these lines :
https://github.com/jfrog/artifactory-user-plugins/blob/master/filestore/nugetFolderMover/nugetFolderMover.groovy#L48
https://github.com/jfrog/artifactory-user-plugins/blob/master/filestore/nugetFolderMover/nugetFolderMover.groovy#L52-L53

seems to generate data ONLY BASED on regex and string split of the fileName

Nuget as a specification, for the format.
This specification force nuget producer to actually respect a certain Folder Structure and the content of it
For example it generate a file with the extension "nuspec"

Can you please change the extraction of "id" an "version" in order to actually read this file and the tags `id` and `version`
This is kinda weird since your base code here has something speaking about `artifactSpec` and variable are allready name `id` and `version`

The thing is that if i create :
* Toto.nupkg
rename it to Tata.nupkg

i should not be able to push the second one, since the `nuspec` file should use as a public key `id`+ `version`

This plugin has the exact same issue

if my nuspec contains :

```
<id>Abcd</id>
<version>1.2.3-rc2-pouet</version>
```

name the file `foo.bar.kix.v14.3.nupkg`
this plugin create :
`foo > bar > kix > v14.3 > THE FILE`

it should do :
`nugetName > version`
`Abcd` > `1.2.3-rc2-pouet` <== note that there is no extra `v` in the version as the SPEC `version` is SemVer2 ....

Worst part ...
let's say you do
`Foo.nupkg`
`Foo.Bar.nupkg`
`Foo.Kix.nupkg`
`Foo.Bar.Kix.nupkg`
`Foo.Kix.Bar.nupkg`

the layout zill split on the '.' char so you'll have

* Foo/
  * version/
      * Foo.nupkg
  * Bar/
    * version/
      * Foo.Bar.nupkg
    * Kix/
      * version/
        * Foo.Bar.Kix
  * Kix/
      * version/
        * Foo.Kix.nupkg
        * Bar/
          * version /
            * Foo.Kix.Bar.nupkg

Why ?

Please simplify it to something like :
as you see the `version` is "formatted" aside the `id` 
-> Just for the UI
-> Not from file name
-> Not in a folder for one file ... as nuget is ONE FILE ONLY
-> No `v` prefix

* Foo/
  * "Foo Version"
* Foo.Bar/
  * "Foo.Bar Version"
* Foo.Kix/
  * "Foo.Kix Version"
* Foo.Bar.Kix/
  * "Foo.Bar.Kix Version"
* Foo.Kix.Bar/
  * "Foo.Kix.Bar Version"

**Comment by DarthFennec** [comment]
I'm having some trouble understanding the issue. It seems that you're talking about the `nugetFolderMover` plugin, is that correct?

>When this plugin is applied on an existing repository, the result of these lines : [snip] seems to generate data ONLY BASED on regex and string split of the fileName

But, the lines you reference here gather information only by looking at the artifact properties, and they do not look at the file name at all. That's how this plugin works: it looks at the artifact properties, and generates an "appropriate" file path/name for the artifact just using that information, regardless of what the original file path/name was. The `nuget.id` and `nuget.version` properties are automatically set by Artifactory when the file is deployed, and their values are extracted from the package's `nuspec` file, not from the file name.

>Please simplify it to something like : [snip]

The actual folder layout used by the `nugetFolderMover` plugin is whatever the repository is configured to use. If you'd like to change the layout, you can configure your repository to use a custom one. [Documentation](https://www.jfrog.com/confluence/display/RTF/Repository+Layouts)

Please let me know if this has been helpful, or if there's something you'd like to ask or clarify further. Thanks!

**Comment by tebeco** [comment]
> When this plugin is applied on an existing repository, the result of these lines : [snip] seems to generate data ONLY BASED on regex and string split of the fileName

this is the biggest issue
the layout should be native for nuget 
it should probably not be done by a layout plugin
and that plugin should not even once works on filename and regex

**Comment by tebeco** [comment]
> That's how this plugin works: it looks at the artifact properties, and generates an "appropriate" file path/name for the artifact just using that information, regardless of what the original file path/name was

i tried to “draw” the generated tree and the “expected”
it really appli to filename and split on ‘.’
also i could not find the code parsing the nuspec file

but manually rename the file generate a new layout

it means that is use filename and not nuget Specification

**Comment by DarthFennec** [comment]
>the layout should be native for nuget
>it should probably not be done by a layout plugin

Artifactory never enforces layouts, although it does support them for all package types. Some types, such as nuget, allow a layout optionally, but do not require one. Instead, Artifactory identifies and indexes nuget packages based entirely on their `nuget.id` and `nuget.version` properties, regardless of the file name, path, or layout.

This plugin is an optional means of enforcing an arbitrary layout in nuget repositories for organizational purposes. This is useful to browsing humans, and inconsequential to Artifactory or the nuget client. That's why it's a plugin, and not native behavior.

>and that plugin should not even once works on filename and regex

I agree that it shouldn't. And it doesn't. There is not a single regex in the `nugetFolderMover` plugin, and the filename is never used during the generation of the new path. I'm not sure what you're seeing in the source that makes you think otherwise.

>it really appli to filename and split on ‘.’

That's because the layout your nuget repository is configured to use is telling it to do that. As previously stated, you can change the way the paths are generated by editing the layout (or creating a new custom one).

>also i could not find the code parsing the nuspec file

That code is not part of this plugin, it's something Artifactory does natively. When a nuget package is deployed, Artifactory inspects its `nuspec` file and extracts the id and version, and uses those values to create new `nuget.id` and `nuget.version` properties on the artifact. Those properties are what Artifactory looks for when indexing or finding nuget packages, and they are also what this plugin looks at when generating a file path.

>but manually rename the file generate a new layout
>it means that is use filename and not nuget Specification

Hm ... that seems like unexpected behavior. It's definitely supposed to look in the `nuspec`, so if it's actually just parsing the filename, that's a bug in Artifactory, I believe. I'll double-check.

>do you have a free docker image i could use on a personal laptop so i could try to do gif ?
>i remember spinning on docker for windows but it either as for key or was not unlocking nuget

There's an image at `docker.bintray.io/jfrog/artifactory-pro`, but it'll require you to enter a license key. Artifactory will not allow you to use nuget repositories (or user plugins) without a key, unfortunately.

**Comment by tebeco** [comment]
> Artifactory never enforces layouts, although it does support them for all package types. Some types, such as nuget, allow a layout optionally, but do not require one. Instead, Artifactory identifies and indexes nuget packages based entirely on their nuget.id and nuget.version properties, regardless of the file name, path, or layout.

weird as we tried it and it just drop all by default in the root folder and no layout
we asked support that told us this plugin as it was the only way to have a proper layout for nuget 

since the variable seems to match the xml tag in the nuspec i agree with you
still default run does none of that :s

**Comment by tebeco** [comment]
> There's an image at docker.bintray.io/jfrog/artifactory-pro, but it'll require you to enter a license key. Artifactory will not allow you to use nuget repositories (or user plugins) without a key, unfortunately.

can you provide a temporary key ?
i would be able to test on a clean artifactory as in my company i have no right on it expect the push on the repo
i would provide on my sides repro step witch scripts that that plays along with the nuget API


also can you confirm me that whatever the layout i need the url used for the push is always the same and i should not sinamycally generate a push url to get a specific layout ?

**Comment by DarthFennec** [comment]
I just did some manual testing for this. Here were my steps:

- Create a nuget repository.
- Find a valid nuget package (I downloaded `angularjs.1.4.8.nupkg` from the official repo).
- Deploy the package to the root of the nuget repository with name `abcxyz.nupkg`.
- Allow the plugin to run.

After the plugin finished running, it had moved the `abcxyz.nupkg` file to `angularjs/angularjs/angularjs.1.4.8.nupkg`. So, the plugin does appear to be using the `nuspec`, rather than the file name.

<hr>

>weird as we tried it and it just drop all by default in the root folder and no layout

The general expectation is that the client or user or CI tool that deploys an artifact "knows what they're doing" and  will put it in the correct place in the repository, so there's no need to force the deploy to conform to a layout. I don't use nuget much so I don't know how much that generalization applies in this case, but it sounds like the nuget client just dumps artifacts in the root of the repository.

This might still be considered fine, because Artifactory identifies/indexes nuget packages based on their `nuget.id` and `nuget.version` properties (not their names or paths), so even if all the packages were just in the root of the repository, they would still be usable by nuget clients. However, it's a good idea to still have a layout so you can browse packages in the UI effectively.

>we asked support that told us this plugin as it was the only way to have a proper layout for nuget

Yeah, that does seem to be the case. This plugin just scans the repository every ten seconds and moves misplaced packages into the repository's layout. So if you want to force the layout, this plugin is the way to do it.

>can you provide a temporary key ?
>i would be able to test on a clean artifactory as in my company i have no right on it expect the push on the repo

I unfortunately can't help you there. You could try asking your company if there's a spare you can use (we generally recommend that people buy a spare "developer" license to be used in situations like this), or you might be able to request a temporary key from your JFrog Success representative.

>also can you confirm me that whatever the layout i need the url used for the push is always the same and i should not sinamycally generate a push url to get a specific layout ?

If you're using this plugin, it can always just be the same url. That will push packages to the root of the repository, and then the plugin will move them to the correct locations.

**Comment by tebeco** [comment]
I will read this entire issue at work today and see with our team if i can close this issue
It seems that the info you provide me here a really looking good
While all the attempts we made does not seems to behave the same way at all

And our JFrog contact also seems to point that the beahior we are seeing is "normal", in a sense that this is what the plugin is design as of today but will change to match the nuspec

Can you tell me the minimal version required of Artifactory + this plugin in order to have the layout done :
* On push (not a timed deamon) <== may be not possible today 
* Based on Nuspec
* Folder is the full name of the nuget using `id`
* every nuget inside with a simple "string format" for id + version
* if you cannot "string format" then 1 folder per version and a signle fil ein it <== waste of UI

**Comment by tebeco** [comment]
I think i found the issue here ...
Will probably close it, as i appologize i did not see that this was a mono repo to all plugins

**Comment by DarthFennec** [comment]
I see. Yes, `uniqueNugetDeploy` is based on file name. To my knowledge, that plugin must run before the file contents are actually available, because that's the only time it's able to cancel the deploy. For that reason, the file name is the only information it has to work with, so that's the reason it gets used instead of something more concrete.

I'm always happy to help ^^ Thanks for taking the time to ask.

**Comment by tebeco** [comment]
closing this ticket until we do a full check in my company
sorry to have bother you with the wrong plugin

**Comment by tebeco** [comment]
@DarthFennec

we were able to make `uniqueNugetFolder` do what we want but there's lot of weird thing :
* `org` or `orgPath` are mandatory in the Layout ... why ?
* `org` makes no sense at all for nuget, please removes this constrains :s we are lucky that `org` actually match `module`
* the docs about layout is purelly `java` oriented, so it can't be used, please duplicate the docs so that it could be specific to each techno
* when publishing artifact, go to the UI, select it, can you please add the list of each "magic value" artifactory computed per nuget so that we don't have to delete content and try it multiple times :
  * `org`
  * `orgPath`
  * `module`
  * ....




> For that reason, the file name is the only information it has to work with, so that's the reason it gets used instead of something more concrete.
This is a huge blocker, as a filename makes no sense. It makes lots of sense to actually parse the nuspec BEFORE the validation so you can cancel it.
Please remember that a filename is not case sensitive in Windows Operating system

This specific problem will crash for maven, ruby, python etc .... that generates packages.
if for example someone did a typo `Toto.TAta` and rename it to `Toto.Tata` and forget to change the version you will endup with a really bad collision when fecting `toto.tata`

**Comment by DarthFennec** [comment]
Yeah, layouts were designed specifically for Java-centric repository types, like Maven, Gradle, and Ivy. That's why none of the other repository types depend on or require layouts, although you can still use them in order to work with layout-oriented tools like version cleanup and certain plugins.

Anything related to how layouts work isn't really in scope for user plugins, so if you want relaxed constraints or semver support, you're best off opening a ticket on the JFrog JIRA. Personally I like these ideas, since they would make layouts more general to any package type, and therefore more useful.

>when publishing artifact, go to the UI, select it, can you please add the list of each "magic value" artifactory computed per nuget so that we don't have to delete content and try it multiple times

As far as I know, you can click on an artifact in the UI and see its basic layout information, assuming it matches the layout. The field is `Module ID`, and it's in the form `<org>:<module>:<version>`. That doesn't cover all the fields, but the important ones are there. If you need more than that, again, you can create a ticket on the JFrog JIRA.

>It makes lots of sense to actually parse the nuspec BEFORE the validation so you can cancel it.

I agree, the problem is that in general cancellations should happen before a file is uploaded. As far as I'm aware, Artifactory is designed so that a deploy cannot be cancelled once the file actually starts uploading. The only alternative I can think of in the general case is to wait for the upload to complete, and then delete the artifact instead of cancelling it, which is a weird hack that I'm unsure will work in all situations.

If, however, we're specifically talking about `uniqueNugetDeploy` and mis-casing, it might be possible to use the file name and just do a case-insensitive search. I was under the impression that this is already done, but maybe not. In that case, this would be another issue for the JIRA, I'd suspect.

**Comment by tebeco** [comment]
> As far as I know, you can click on an artifact in the UI and see its basic layout information, assuming it matches the layout. The field is Module ID, and it's in the form <org>:<module>:<version>. That doesn't cover all the fields, but the important ones are there

well i did not see `org` or `orgPath` and at least one of them is mandatory so it’s kinda critical
Also it would feel better if the displayed field had the exact same name of the one from the layout and probably grouped together in a table named for that ;)

**Comment by tebeco** [comment]
> I agree, the problem is that in general cancellations should happen before a file is uploaded. As far as I'm aware, Artifactory is designed so that a deploy cannot be cancelled once the file actually starts uploading. 

i’m sorry i laughted ^^ so corrupted file are ok ?
this is reaaaaaaally weird
how can a validation be done before upload server side ?

**Comment by DarthFennec** [comment]
Artifactory manages the artifacts you give it, and for the most part it doesn't really care much what the content of an artifact actually is. If you really want to upload a corrupted file, it doesn't stop you. There are some situations where it does reject artifacts based on their contents (for example, if a Maven repository has the "Suppress POM Consistency Checks" box unchecked, it will reject POM files if their file paths don't match their internal data), but situations like this are special-case, and from what I can tell this behavior is not available in general to user plugins.

In a user plugin, we get two hooks for file uploads: `beforeCreate` and `afterCreate`. According to the documentation, a `CancelException` can be thrown in `beforeCreate` to cancel the upload, but no such claim is made about `afterCreate`. I haven't tried throwing this exception in `afterCreate`, so I don't know for certain that it doesn't work, but I doubt it would. You're welcome to try if you'd like, but I'd be (pleasantly) surprised if it did the trick.

As for `beforeCreate`, it would be impossible to validate the file contents, because at that point in the deployment process, the file hasn't been uploaded yet. Artifactory can't validate a file that it doesn't have.

**Comment by markgalpin** [comment]
To be clear there is some additional validation done on the artifact post-upload, it is just not accessible for modification via a plugin.  For example if a checksum is provided by the client (or via the checksum header) artifactory will calculate the checksum of the received object and return a 409 if it doesn't match what the headers told it it should receive.  But mostly we're not validating the quality of object you upload, just that you sent us what you said you would.  On a technology by technology basis there may be other checks, such as the maven check @DarthFennec mentioned earlier, but mostly we assume you are uploading what you intend to upload.  If the package is not well-formed as a NuGet package we may accept it (depending on how it is provided: we always honor the client API if it has specific validation expectations and you are using it), but not include it in the NuGet index for retrieval by NuGet.

**Comment by tebeco** [comment]
> On a technology by technology basis there may be other checks, such as the maven check @DarthFennec mentioned earlier, but mostly we assume you are uploading what you intend to upload.


is there a repository technology that allow you freely to complitly obliterate an existing artifact ?

speaking of technology, Filesystem works by filename and ask you what to do unless you `—force`

Artifactory does artifacts not files, files are just a representation for you server, it could be blob.
also when pushing an artifact it’s just a bytestream so again no no no this is not technology by technology
i mean we could use an ftp server for that and use an ftp client for the viewer 

using a dedicated, payed, repository content manager is here stuff like checking 
if packages already exists to avoid mitigation
if the file is corrupted (can it unzip it and does it respect the specs ... because you implements the specs, this is why nuget repository is not free right ?)
if the package is signed to avoid injected dll if apikey is compromised (this is a hint, please note that nuget is moving toward nuget signing and removing apikey as an auth system for publication)

these checks are the foundation for any repository regardless of their respective technology

**Comment by DarthFennec** [comment]
>is there a repository technology that allow you freely to complitly obliterate an existing artifact ?

You can delete any artifact from any repository. Is that what you mean by "obliterate"?

>using a dedicated, payed, repository content manager is here stuff like checking if packages already exists to avoid mitigation
>if the file is corrupted (can it unzip it and does it respect the specs ... because you implements the specs, this is why nuget repository is not free right ?)
>
>these checks are the foundation for any repository regardless of their respective technology

I think there's a misunderstanding here. As far as I'm aware, Artifactory does follow the nuget specs and API properly when using a nuget repository. As long as you use the `api/nuget/<reponame>/` endpoint to access a nuget repository, Artifactory will ensure the package is not corrupted and the specs are respected, and the package doesn't already exist in the repository. However, in the general case (or if you use the generic interfaces to upload an artifact) it doesn't check these things, and also it doesn't check for duplicates between separate repositories (the problem `uniqueNugetDeploy` is designed to solve).

I hope that clarifies things a little.

**Comment by tebeco** [comment]
> You can delete any artifact from any repository. Is that what you mean by "obliterate"?

i mean push twice same stuff based on a name and not on the spec that identify it

**Comment by DarthFennec** [comment]
>i mean push twice same stuff based on a name and not on the spec that identify it

If you deploy two different artifacts with the same path/name to the same repository, the second one will overwrite the first, yes. Unless of course you don't have delete permissions for that path, in which case Artifactory will reject the second deploy. Is this what you're asking about?

**Comment by DarthFennec** [comment]
Closing this for now. Feel free to reopen if you have any further questions on this.
